### ISSUE TITLE
[BUG] `sys_alloc_aligned` does not seem to impose a minimal alignment

### ISSUE THREAD

**Issue by weikengchen** [risc0/risc0#1209]
## Bug Report

There is a very confusing line of code in the exported syscall of `sys_alloc_aligned`, which is behind RISC Zero's memory allocator.

Recall that the bump memory allocator is here:
https://github.com/risc0/risc0/blob/main/risc0/zkvm/platform/src/syscall.rs#L651

```rust
#[cfg(feature = "export-syscalls")]
#[no_mangle]
pub unsafe extern "C" fn sys_alloc_aligned(bytes: usize, align: usize) -> *mut u8 {
    extern "C" {
        static _end: u8;
    }
    static mut HEAP_POS: usize = 0;
    let mut heap_pos = unsafe { HEAP_POS };
    if heap_pos == 0 {
        heap_pos = unsafe { (&_end) as *const u8 as usize };
    }

    let offset = heap_pos & (align - 1);
    if offset != 0 {
        heap_pos += align - offset;
    }

    // Ensure all allocations are minimally aligned to a word boundary.
    let align = usize::min(align, WORD_SIZE);

    let ptr = heap_pos as *mut u8;
    heap_pos += bytes;
    
    if crate::memory::SYSTEM.start() < heap_pos {
        const MSG: &[u8] = "Out of memory!".as_bytes();
        unsafe { sys_panic(MSG.as_ptr(), MSG.len()) };
    }

    unsafe { HEAP_POS = heap_pos };
    ptr
}
```

The problem rests on the "align".

- First of all, is it `min` or `max`? If we want to make sure that the alignment is at least WORD_SIZE, it should be max.
- Second, "let align" has no meaningful purpose because the `offset` and `heap_pos` have been updated before "let align = usize::min(align, WORD_SIZE);"? `align` is never used after this.

## Expected behavior
Expected `min` to be `max` and the align line be put up before the computation of offset.

**Comment by SchmErik** [comment]
Thanks for filing this. So it looks like we've been aligning to what has been passed in. IIUC, this function is meant to be called from rust's `alloc` crate and `align` is supposed to be a power of 2 that represents how memory should be aligned as a part of `alloc::Layout`. I'm wondering if we could simply remove the max/min and pass on the responsibility to the caller of the function. @flaub do you have any thoughts or info on this?

**Comment by weikengchen** [comment]
It seems that Rust decides the `alloc::Layout` based on that data type, so byte would be byte, u32 would be u32, unless additional attributes are present that modifies the alignment request. 

The main problem would be: is there a very commonly occurring situation where not matching up the alignment---by the way, in the heap---would cause significant slowdown or problems?

**Comment by SchmErik** [comment]
> It seems that Rust decides the `alloc::Layout` based on that data type, so byte would be byte, u32 would be u32, unless additional attributes are present that modifies the alignment request.

I'm sorry! I was misunderstanding `alloc::Layout` and you're right about this.

**Comment by SchmErik** [comment]
Sorry for the delayed response - I realized that we rely on aligning certain structs to page boundary and in general, I think it's a good idea to honor the requested alignment which is a power of 2. So I changed the logic to align to either WORD_SIZE or the input which (within the rust ecosystem) must be a power of two. PR available here: #1239
